src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts

```diff
@@ -0,0 +1,106 @@
+import { RequestHandler } from 'express'
+import { ParamsDictionary } from 'express-serve-static-core'
+import { StatusCodes } from 'http-status-codes'
+import JSONStream from 'JSONStream'
+
+import { createLoggerWithLabel } from '../../../../config/logger'
+import { IPopulatedForm } from '../../../../types'
+import { createReqMeta } from '../../../utils/request'
+
+import { getSubmissionCursor } from './encrypt-submission.service'
+
+const logger = createLoggerWithLabel(module)
+
+type WithForm<T> = T & {
+  form: IPopulatedForm
+}
+
+/**
+ * Handler for GET /:formId([a-fA-F0-9]{24})/adminform/submissions/download
+ *
+ * @returns 200 with stream of encrypted responses
+ * @returns 400 if req.query.startDate or req.query.endDate is malformed
+ * @returns 500 if any errors occurs in stream pipeline
+ */
+export const handleStreamEncryptedResponses: RequestHandler<
+  ParamsDictionary,
+  unknown,
+  unknown,
+  { startDate?: string; endDate?: string }
+> = async function (req, res) {
+  const { startDate, endDate } = req.query
+
+  // TODO (#42): Remove typecast once app has migrated away from middlewares.
+  const formId = (req as WithForm<typeof req>).form._id
+
+  const cursorResult = getSubmissionCursor(formId, {
+    startDate,
+    endDate,
+  })
+
+  if (cursorResult.isErr()) {
+    return res.status(StatusCodes.BAD_REQUEST).json({
+      message: 'Malformed date parameter',
+    })
+  }
+
+  const cursor = cursorResult.value
+
+  const logMeta = {
+    action: 'handleStreamEncryptedResponses',
+    ...createReqMeta(req),
+    formId,
+  }
+
+  cursor
+    .on('error', (error) => {
+      logger.error({
+        message: 'Error streaming submissions from database',
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
+        message: 'Error retrieving from database.',
+      })
+    })
+    // If you call JSONStream.stringify(false) the elements will only be
+    // seperated by a newline.
+    .pipe(JSONStream.stringify(false))
+    .on('error', (error) => {
+      logger.error({
+        message: 'Error converting submissions to JSON',
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
+        message: 'Error converting submissions to JSON',
+      })
+    })
+    .pipe(res.type('application/x-ndjson'))
+    .on('error', (error) => {
+      logger.error({
+        message: 'Error writing submissions to HTTP stream',
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
+        message: 'Error writing submissions to HTTP stream',
+      })
+    })
+    .on('end', () => {
+      logger.info({
+        message: 'Stream encrypted responses ended',
+        meta: logMeta,
+      })
+
+      return res.end()
+    })
+    .on('close', () => {
+      logger.info({
+        message: 'Stream encrypted responses closed',
+        meta: logMeta,
+      })
+
+      return res.end()
```

Comment:
what's the difference between `end` and `close`? we previously didn't have `close` right? might there be errors if `res.end()` is called twice?

Reply:
That's a good question. From my refactoring and the additional tests, the `end` event was actually never emitted! So I added the close event handler. Maybe I'll remove the end handler since the close event should always be emitted at the end

Reply:
I don't think you should use [close](https://nodejs.org/api/stream.html#stream_event_end_1) since our code simply receives the stream object from mongoose, but isn't actually in control of the stream. I believe you'll also end up with 5XX errors if mongoose decides to close the stream and `res.end()` ends up being called twice.

Reply:
As discussed, end is not called, removed and use close instead.

Will create issue to allow for frontend to also stop waiting once the pipe is closed on the backend

Reply:
this should already be compatible with the frontend, so feel free to merge.